tree/timer.go

```diff
@@ -0,0 +1,105 @@
+package tree
+
+import (
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-backoff/policy"
+	"github.com/libp2p/go-libp2p-core/backoff"
+)
+
+func NewBackoffTreeTimer(parent backoff.BackoffTimer, policy policy.BackoffPolicy) *BackoffTreeTimer {
+	return &BackoffTreeTimer{
+		parent:   parent,
+		state:    policy.NewBackoffState(),
+		children: map[string]*BackoffTreeTimer{},
+	}
+}
+
+type BackoffTreeTimer struct {
+	parent backoff.BackoffTimer
+	sync.Mutex
+	clk      sync.Mutex // lock for children
+	children map[string]*BackoffTreeTimer
+	slk      sync.Mutex // lock for state
+	state    policy.BackoffState
+}
+
+func (t *BackoffTreeTimer) Subtimer(childName string, policy policy.BackoffPolicy) *BackoffTreeTimer {
+	t.clk.Lock()
+	defer t.clk.Unlock()
+	if child := t.children[childName]; child != nil {
+		return child
+	} else {
+		d := NewBackoffTreeTimer(t, policy)
+		t.children[childName] = d
+		return d
+	}
+}
+
+var BackoffGCInterval = time.Minute
+
+func (t *BackoffTreeTimer) StartGC() {
```

Comment:
a task like this should probably run within a `context` so that it can be completed on shutdown.